DllImportGenerator/DllImportGenerator/StubCodeContext.cs

```diff
@@ -46,7 +46,14 @@ public enum Stage
             /// <summary>
             /// Keep alive any managed objects that need to stay alive across the call.
             /// </summary>
-            KeepAlive
+            KeepAlive,
+
+            /// <summary>
+            /// Convert native data to managed data
+            /// where native values will leak if we
+            /// fail to unmarshal, but do run cleanup.
+            /// </summary>
+            LeakSafeUnmarshal
```

Comment:
I think this is getting too hard to understand. Why is this different from Unmarshal and Cleanup?

Reply:
It needs to be different from Unmarshal because SafeHandle unmarshalling has to happen in the `finally` block instead of in the `try` block (once we add the try-finally to the emitted code). It has to happen in the finally block since the handles will leak if an exception is thrown (copying the behavior from CoreCLR and CoreRT).

Additionally, the SafeHandle unmarshalling conceptually isn't cleanup. We aren't cleaning up native resources here, we're unmarshalling the handle into a SafeHandle-derived object. Since we have the ability to introduce new stages and we don't have to overload the meaning of "cleanup", I chose to introduce a new stage.

Reply:
@jkoritzinsky I agree with this argument but I think where this is occurring might be more beneficial than such a specific name. Perhaps `FinallyBlock`? I recall the debate about `GuaranteedCleanup` not being accurate since an exception can throw in a `finally` - perhaps letting that slide here is okay because if an exception is thrown couldn't we consider that an ill-formed `GuaranteedCleanup` section?

This is similar to my feedback about `KeepAlive` which might be easier to call `PostInvoke`. I don't recall the argument for why stating it as `KeepAlive` was appropriate but do recall thinking it was valid.

Reply:
I'd be ok calling it `GuaranteedUnmarshal` and saying that an exception thrown during execution of this section is ill-formed. I'm not a fan of calling this section a cleanup section since it's doing unmarshalling and the Cleanup section is emitted in the same finally block.

IIRC with `KeepAlive` we just kind of left it as-is. I think the reasoning was that we couldn't think of anything else that would go in that section, and that we would rename it once something else fit to go there.